src/Hosting/Abstractions/src/MetricsPrototype/DefaultMeterFactory.cs

```diff
@@ -0,0 +1,107 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.Metrics;
+using System.Linq;
+using Microsoft.Extensions.Metrics;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Hosting.Abstractions.MetricsPrototype;
+
+internal sealed class DefaultMeterFactory : IMeterFactory
+{
+    private readonly IOptions<MetricsOptions> _options;
+    private readonly IMeterRegistry _meterRegistry;
+    private readonly Dictionary<MeterKey, Meter> _meters = new Dictionary<MeterKey, Meter>();
+
+    public DefaultMeterFactory(IOptions<MetricsOptions> options, IMeterRegistry meterRegistry)
+    {
+        _options = options;
+        _meterRegistry = meterRegistry;
+    }
+
+    public Meter CreateMeter(string name)
+    {
+        return CreateMeterCore(name, version: null, defaultTags: null);
+    }
+
+    public Meter CreateMeter(MeterOptions options)
+    {
+        return CreateMeterCore(options.Name, options.Version, options.DefaultTags);
+    }
+
+    private Meter CreateMeterCore(string name, string? version, IList<KeyValuePair<string, object?>>? defaultTags)
+    {
+        var tags = defaultTags?.ToArray();
+        if (tags != null)
+        {
+            Array.Sort(tags, (t1, t2) => string.Compare(t1.Key, t2.Key, StringComparison.Ordinal));
+        }
+        var key = new MeterKey(name, version, tags);
+
+        if (_meters.TryGetValue(key, out var meter))
+        {
+            return meter;
+        }
+
+        // TODO: Configure meter with default tags.
+        meter = new Meter(name, version);
+        _meters[key] = meter;
+        _meterRegistry.Add(meter);
+
+        return meter;
```

Comment:
The factory returns an existing meter instance if the same name/version/tags are passed to it. Thoughts?

Reply:
I think it should, and similarly Meter probably needs to be changed so it returns the same instrument if called with the same parameters. If we don't have that behavior it seems too easy to inadvertently create large memory leaks from people creating Meters and instruments inside of scoped and transient lifetime DI services.